File: Data Structures & Algorithms/count-number-of-islands/submission-5.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return []

        res = 0
        m = len(grid)
        n = len(grid[0])

        def dfs(r, c):
            grid[r][c] = '0'
            for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                nr, nc = r + dr, c + dc
                if not 0 <= nr < m or not 0 <= nc < n or grid[nr][nc] == '0':
                    continue
                dfs(nr, nc)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    dfs(i, j)
                    res += 1

        return res

File: Data Structures & Algorithms/max-area-of-island/submission-5.java
class Solution {
    private static final int[][] directions = {{1, 0}, {-1, 0}, {0, -1}, {0, 1}};

    public int maxAreaOfIsland(int[][] grid) {
        int ROWS = grid.length;
        int COLS = grid[0].length;

        int maxArea = 0;

        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                if (grid[i][j] == 1) {
                    int area = dfs(grid, i, j);
                    maxArea = Math.max(maxArea, area);
                }
            }
        }

        return maxArea;
    }

    private int dfs(int[][] grid, int i, int j) {
        int area = 0;
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length) {
            return area;
        }

        if (grid[i][j] == 1) {
            area++;
            grid[i][j] = 0;
            for (int[] d: directions) {
                area += dfs(grid, i + d[0], j + d[1]);
            }
        }

        return area;
    }
}
